load_continuous_data takes slice/list as an argument

crheller opened the pull request.

bburan commented on nems_lbhb/io.py:
```diff
@@ -176,15 +177,33 @@ def _get_baphy_parameters(self):
-            with tar_fh.extractfile(str(full_filename)) as fh:
-                return load_continuous_openephys(fh)
+        # get filenames
+        all_files = os.listdir(self.openephys_folder)
```
Lines 187 and 188 can be replaced by:
`data_files = self.openephys_folder.glob('*CH*')`

bburan replied:
Also, will this work with tarfiles?

bburan replied:
If the files are only inside the tarfile, you can open the tarfile and then use the `getmembers` method to get a list of the filenames stored within.

bburan commented on nems_lbhb/io.py:
```diff
@@ -176,15 +177,33 @@ def _get_baphy_parameters(self):
+        selected_data = np.take(data_files, idx)
+        # TODO add channel remapping?
+        
+        for i, filename in enumerate(selected_data):
```
Better to create a list containing the data then do one concatenate at the end:
```
continuous_data = []
for ...
    continuous_data.append(data['data'][np.newaxis, :])

continuous_data = np.concatenate(continuous_data, axis=0)
```

crheller commented:
I think I've addressed the changes you mentioned. Finding the files using getmembers() or getnames() is quite slow, as is loading from the tarfile every time. I added a case to speed up loading if the data has already been extracted. Can you think of a way to speed up the search for filenames?

crheller commented:
Also, maybe we want to think about caching the files when we extract them? Or will that take up too much disk space on elephant?

bburan commented:
Isn't the Recording cached once it's created? If so, that may be the best place for caching.

Brad

________________________________
From: crheller [notifications@github.com]
Sent: Wednesday, June 26, 2019 10:21 AM
To: LBHB/nems_db
Cc: Brad Buran; Review requested
Subject: Re: [LBHB/nems_db] load_continuous_data takes slice/list as an argument (#25)


Also, maybe we want to think about caching the files when we extract them? Or will that take up too much disk space on elephant?

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub<https://github.com/LBHB/nems_db/pull/25?email_source=notifications&email_token=AACPSQQCX2QTXIXEL6DWSALP4OQRBA5CNFSM4H3NIYUKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYUHMFY#issuecomment-505968151>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AACPSQS6PRL2BJKEEK7BAS3P4OQRBANCNFSM4H3NIYUA>.

crheller commented:
Yeah, that's a fair point. Otherwise, looking okay at this point?

bburan commented on nems_lbhb/io.py:
```diff
@@ -176,15 +177,46 @@ def _get_baphy_parameters(self):
+                    with tar_fh.extractfile(str(full_filename)) as fh:
+                        data = load_continuous_openephys(fh)
+                        continuous_data.append(data['data'][np.newaxis, :])
+                        tar_fh.close()
```
You do not need to explicitly close `tar_fh` or `fh`. These are done by the with statement as part of the cleanup when exiting the block.

bburan commented on nems_lbhb/io.py:
```diff
@@ -176,15 +177,46 @@ def _get_baphy_parameters(self):
+
+        continuous_data = np.concatenate(continuous_data, axis=0)
+
+        # TODO return timestamps and other meta data?
```
Timestamps are already handled by another method on this class. Metadata would also be useful, but I would handle this as a separate method as well (unless, for performance reasons, it makes sense to return as part of this function).

bburan commented on nems_lbhb/io.py:
```diff
@@ -176,15 +177,46 @@ def _get_baphy_parameters(self):
+                continuous_data.append(data['data'][np.newaxis, :])
+            else:
+                with tarfile.open(self.openephys_tarfile, 'r:gz') as tar_fh:
+                    log.info("Extracting / loading {}...".format(filename))
```
It's good to get in the habit of (for logging messages), doing the following:

`log.info("Extracting / loading %s ...", filename)`

In this approach, the string only gets formatted if the logging is enabled. If it's not enabled, no formatting is performed. In the approach you use, the string is always formatted regardless of whether logging is enabled or not. In this particular situation, the "overhead" is minimal. However, if you're doing this formatting inside one of the functions that's called during the fitting process, the overhead can add up.
